File: legacy/src/vpn.ts
import { existsSync } from 'fs';
import { join } from 'path';
import { execFileSync } from 'child_process';
import { get } from 'https';

const NORDVPN_EXE = 'nordvpn.exe';
const IP_CHECK_URL = 'https://api.ipify.org';
const IP_CHECK_TIMEOUT = 8000;
const RECONNECT_WAIT_MS = 3000;
const IP_POLL_INTERVAL_MS = 1500;
const IP_POLL_MAX_ATTEMPTS = 20;

// Códigos de país para NordVPN (2 letras, minúsculas)
// Variedad de países europeos y latinoamericanos para evitar patrones
const VPNPAISES = [
  'Spain',
  'France',
  'Germany',
  'Italy',
  'Portugal',
  'switzerland',
  'Belgium',
  'India',
  'Sweden', 
  'Ireland',
  'Brazil', 
  'Mexico', 
  'Albania', 
  'Mexico',
  'Poland',
  'United States',
  'South Korea',
  'Hong Kong',
  'Singapore',
  'Austria',
  'Norway'
];

function paisAleatorio(): string {
  return VPNPAISES[Math.floor(Math.random() * VPNPAISES.length)]!;
}

function resolveExe(nordvpnPath: string): string {
  return join(nordvpnPath, NORDVPN_EXE);
}

export function nordvpnDisponible(nordvpnPath: string): boolean {
  return existsSync(resolveExe(nordvpnPath));
}

async function obtenerIpPublica(): Promise<string | null> {
  return new Promise((resolve) => {
    const timer = setTimeout(() => resolve(null), IP_CHECK_TIMEOUT);
    get(IP_CHECK_URL, (res) => {
      let data = '';
      res.on('data', (chunk) => { data += chunk; });
      res.on('end', () => {
        clearTimeout(timer);
        resolve(data.trim() || null);
      });
    }).on('error', () => {
      clearTimeout(timer);
      resolve(null);
    });
  });
}

async function esperarCambioIp(ipAnterior: string): Promise<void> {
  for (let i = 0; i < IP_POLL_MAX_ATTEMPTS; i++) {
    await new Promise(r => setTimeout(r, IP_POLL_INTERVAL_MS));
    const ipNueva = await obtenerIpPublica();
    if (ipNueva && ipNueva !== ipAnterior) return;
  }
}

export async function rotarVpn(nordvpnPath: string, onStatus?: (msg: string) => void): Promise<void> {
  const exe = resolveExe(nordvpnPath);

  onStatus?.('VPN: obteniendo IP actual...');
  const ipAnterior = await obtenerIpPublica();

  onStatus?.('VPN: desconectando...');
  try {
    execFileSync(exe, ['-d'], { timeout: 10000 });
  } catch {}

  await new Promise(r => setTimeout(r, RECONNECT_WAIT_MS));

  onStatus?.('VPN: conectando a nuevo servidor...');
  try {

    execFileSync(exe, ['-c', '-g', paisAleatorio()], { timeout: 15000 });
  } catch {}

  if (ipAnterior) {
    onStatus?.('VPN: esperando nueva IP...');
    await esperarCambioIp(ipAnterior);
  } else {
    await new Promise(r => setTimeout(r, RECONNECT_WAIT_MS));
  }

  onStatus?.('VPN: rotacion completada');
}

File: biz-scraping/src/vpn.rs
use std::path::{Path, PathBuf};

use tokio::process::Command;

use crate::verboser::Verboser;

const NORDVPN_EXE: &str = "nordvpn.exe";

const VPN_COUNTRIES: &[&str] = &[
    "Spain",
    "France",
    "Germany",
    "Italy",
    "Portugal",
    "switzerland",
    "Belgium",
    "India",
    "Sweden",
    "Ireland",
    "Brazil",
    "Mexico",
    "Albania",
    "Mexico",
    "Poland",
    "United States",
    "South Korea",
    "Hong Kong",
    "Singapore",
    "Austria",
    "Norway",
];

fn random_country() -> &'static str {
    let idx = rand::random::<usize>() % VPN_COUNTRIES.len();
    VPN_COUNTRIES[idx]
}

fn resolve_exe(nordvpn_path: &Path) -> PathBuf {
    nordvpn_path.join(NORDVPN_EXE)
}

pub fn nordvpn_available(nordvpn_path: &Path) -> bool {
    resolve_exe(nordvpn_path).exists()
}

pub async fn rotate_vpn(nordvpn_path: &Path, verboser: &impl Verboser) {
    verboser.vpn_rotating();
    let exe = resolve_exe(nordvpn_path);

    let _ = Command::new(&exe).arg("-d").kill_on_drop(true).status().await;

    let _ = Command::new(&exe)
        .args(["-c", "-g", random_country()])
        .kill_on_drop(true)
        .status()
        .await;

    verboser.vpn_rotated();
}
